cdc: use `cdc` schema extension for storing/reading cdc metadata

piodul opened the pull request:
Currently, information on what cdc options are enabled in a table - cdc metadata in short - is stored in two places:

- In `cdc` column of the `system_schema.scylla_tables`,
- In a `cdc` schema extension.

The former is used as a source of truth, i.e. a node reads cdc metadata from that column, while the latter is used for cosmetic purposes (e.g. cqlsh displays info on cdc based on this extension) and is only written, but never read by the node.

Introducing the `cdc` column to `scylla_tables` made the logic of schema agreement more complicated. As a first step of removing this column, this PR makes the `cdc` schema extension as the "source of truth" - a node will from now on read cdc metadata from that extension.

The `cdc` column will be deprecated and removed in subsequent releases, but it is left for now and will still be written to in order not to break the logic of schema agreement.

Refs: #5737
Tests: unit(dev), 2-node cluster upgrade under write load to a cdc-enabled table

jul-stas commented on db/extensions.hh:
```diff
@@ -100,6 +100,26 @@ public:
+    template<typename Extension>
+    void add_schema_extension(sstring w) {
+        add_schema_extension(std::move(w), [] (db::extensions::schema_ext_config cfg) {
+            return std::visit([] (auto v) {
```
Maybe universal reference would work here? (I'm absolutely not insisting)

haaawk replied:
Would Extension v work here?

piodul replied:
@haaawk If you are talking about changing `auto v` to `Extension v`, then no, it wouldn't work. Notice the `std::visit` call - `schema_ext_config` is a typedef for a `std::variant`. Using a generic lambda here allows to construct `Extension` using a constructor overload appropriate for each case of the variant.

haaawk replied:
Ok. Please ignore my comment then.

haaawk commented on db/schema_tables.cc:
```diff
@@ -60,7 +60,6 @@
 #include "mutation_query.hh"
 #include "system_keyspace.hh"
 #include "cql3/cql3_type.hh"
-#include "cdc/log.hh"
```
It's not ok to break codebase with one commit and fix with another. Changes have to be bisectable - every commit has to build correctly and pass all the tests. Please rearrange the patches in such a way to keep every commit build and pass tests.

haaawk replied:
Probably it's enough to apply this patch after switching to using extension.

haaawk commented on cql3/statements/cf_prop_defs.hh:
```diff
@@ -89,11 +89,13 @@ public:
     static constexpr int32_t DEFAULT_MAX_INDEX_INTERVAL = 2048;
 private:
     mutable std::optional<sstables::compaction_strategy_type> _compaction_strategy_class;
+    mutable std::optional<schema::extensions_map> _schema_extensions;
```
Nit: This is more caching of schema_extension than lazy initialization. This patch does not change the time schema extensions will be created. I understand why you're doing that but current commit message is a bit confusing.

piodul replied:
I updated the commit description to name it as "caching", not "lazy initialization".

haaawk commented on cdc/cdc_extension.hh:
```diff
@@ -20,6 +20,7 @@
 
 #pragma once
 
+#include "schema.hh"
```
Please add this in patch that introduces the dependency on schema_extension. Unless it's not one of the patches of this PR.

haaawk commented on test/boost/cdc_test.cc:
```diff
@@ -23,6 +23,7 @@
 #include <string>
 
 #include "cdc/log.hh"
+#include "cdc/cdc_extension.hh"
```
Please merge this patch into the one that breaks the tests and introduces the need for the fixes included in this patch.

nyh replied:
What about this (and potentially other bisectability issues)?

piodul replied:
Those issues now should be resolved. For every commit in this patch, I checked that it compiles and makes unit tests pass. Before uploading the most recent version, I did the same.

haaawk commented:
Looks good to me in general. I left some minor comments. The only major issue is that individual commits don't compile or pass tests (unless I'm wrong) and that's something we shouldn't do. It's important to keep every commit in the history building and passing tests. That way we can use bisect to search for bugs. Please rearrange the commits to fulfil this requirement. Apart from that, good job @piodul!

piodul commented:
v2:

- Rebased onto recent master,
- Rearranged and squashed some commits so that each of them compiles and has passing unit tests,
- Updated description of some of the commits.

piodul commented:
v3: got rid of the commit that added the `schema.hh` import, and merged it into the first commit that requires that import

nyh commented on cql3/statements/cf_prop_defs.cc:
```diff
@@ -188,6 +188,27 @@ std::optional<std::map<sstring, sstring>> cf_prop_defs::get_cdc_options() const
+const schema::extensions_map& cf_prop_defs::get_schema_extensions(const db::extensions& exts) const {
+    if (_schema_extensions) {
+        return *_schema_extensions;
+    }
```
I don't understand this - this function has a parameter, and yet the cached response doesn't depend on this parameter... If you call this function twice with different parameters, the second call will return wrong results!
Maybe you don't plan to call it twice with two different parameters (I only see one call in this patch, I don't know what will come next), but this is still a dangerous function to have around and who knows who will call it in future code. There isn't even a comment warning about not doing that (though I think in this case code changes will be more prudent than a comment).

piodul replied:
That's a good point, I see why it is seems dangerous to leave it like that. The extensions come from the the `database` object, which is shard-local. Those extensions shouldn't change during the execution.

I changed `get_schema_extensions` and `validate` methods to take `database` instead of `extensions` as an argument.

elcallio replied:
Imho this is still not very kosher. Caching like this would be ok if database was a (non-changing) member, but here it is an arbitrary argument. 
If you want a cached `_schema_extensions` it should be generated as such at some point where the code flow guarantees that db is the same. 
And yes, I realize that who database is, is invariant in actual runtime, but remember all the talk of decoupling, multi-tenancy and moreover: testability. 
You technically only need the extensions once, so calculate them uncached, but keep the return value, and make `get_cdcoptions` take this extension object instead.

piodul replied:
I see. I changed it so that it constructs `schema::extensions_map` outside, and passes it to methods of `cf_prop_defs` as a parameter.

nyh commented on cql3/statements/cf_prop_defs.cc:
```diff
@@ -129,11 +128,9 @@ void cf_prop_defs::validate(const db::extensions& exts) const {
-        cdc::options opts(*cdc_options);
-    }
+    // This forces construction of cdc options, if they are present in extensions map
+    // Constructor throws if options are not valid
```
I don't see a throw in get_cdc_options() below, what throws - to_options()?

piodul replied:
`cdc::options` have a constructor that throws when incorrect options are passed to it. The code above is no longer relevant in the new version, because the options are constructed in `cdc_extension`.

nyh commented on schema.cc:
```diff
@@ -1113,6 +1114,14 @@ schema_ptr schema_builder::build() {
     return make_lw_shared<schema>(schema(new_raw, _view_info));
 }
 
+void schema_builder::build_properties_derived_from_extensions() {
```
This is a 100% CDC-only method. Does it need to be in schema.cc/hh?
Moreover, why do we even need the "_cdc_options" schema member any more, if it directly calculated from the new "schema extension"?

piodul replied:
Changed _cdc_options field to a cdc_options() method.

piodul commented:
v2:

- `cdc::options` are now constructed inside `cdc_extension`,
- The `get_schema_extensions` method now receives a reference to a `database` object,
- `schema` now no longer has a `_cdc_options` field. In `cdc_options()` method, it now returns a reference to `cdc::options` that are situated in schema's `cdc_extension`.

elcallio commented on cql3/statements/cf_prop_defs.cc:
```diff
@@ -263,33 +285,19 @@ void cf_prop_defs::apply_to_builder(schema_builder& builder, const database& db)
-        auto opts = cdc::options(*cdc_options);
-        if (opts.enabled() && !db.features().cluster_supports_cdc()) {
+        if (cdc_options->enabled() && !db.features().cluster_supports_cdc()) {
             throw exceptions::configuration_exception("CDC not supported by the cluster");
```
Shouldn't this be in `validate`?

piodul replied:
Moved it to `validate`.

elcallio commented on db/extensions.hh:
```diff
@@ -100,6 +100,26 @@ public:
+     */
+    template<typename Extension>
+    void add_schema_extension() {
+        add_schema_extension<Extension>(Extension::NAME);
```
This seems superfluous, and also assumes magic members, which is confusing when reading the call site. 
Also, proper stl-esque way to do such a thing would be to have a namespace level
```
template<typename T>
struct extension_name_of {
     sstring operator()() const {
        return T::name;
     }
};

// and specialize
template<>
struct extension_name_for<PigletExtension> {
   sstring operator()() const { return ::to_string(time()); }
};
```

But again, superfluous.

piodul replied:
I see. I removed it.

elcallio commented on main.cc:
```diff
@@ -433,12 +433,8 @@ int main(int ac, char** av) {
-        return std::visit([](auto v) { return ::make_shared<cdc::cdc_extension>(v); }, cfg);
-    });
+    ext->add_schema_extension<alternator::tags_extension>();
+    ext->add_schema_extension<cdc::cdc_extension>();
```
Just use the constants as arguments. No magic members.

piodul commented:
v3:

- extensions are no longer cached inside `cf_prop_defs`
- CDC cluster support is now checked in `validate`, not `apply_to_builder`
- removed unnecessary overload of `extension::add_schema_extension`

piodul commented:
@nyh ping

nyh commented on cql3/statements/cf_prop_defs.cc:
```diff
@@ -78,7 +78,7 @@ const sstring cf_prop_defs::COMPACTION_STRATEGY_CLASS_KEY = "class";
 const sstring cf_prop_defs::COMPACTION_ENABLED_KEY = "enabled";
 
-void cf_prop_defs::validate(const db::extensions& exts) const {
+void cf_prop_defs::validate(const database& db) const {
```
I'm sorry, I didn't understand this patch. Why is it important that the "extensions is taken from the global database"? In a previous version of this patch, you had fishy caching here, so it was dangerous if called twice with different extensions, but I think you don't have this caching any more (at least not in the form you had it). So why do we need this patch?

piodul replied:
Ok, I see, sorry, I didn't update the commit description. This patch was useful in resolving one of the review comments ([this comment](https://github.com/scylladb/scylla/pull/5908#discussion_r387513500), fixed by the [next commit](https://github.com/scylladb/scylla/pull/5908/commits/0d4cc56d981af19e05732fb5ffb03f77e72d89b7)).

I updated the description.

nyh commented:
I'll commit this series. I think things could have been prettier. There's still too much messing with "extensions" in various pieces of code that doesn't do anything with extensions, and I didn't understand at all the patch which add "db" to validate, and then apparently there is a second patch which also adds extensions to it. Please do try to find some less stressed period and clean this stuff up a bit.

haaawk commented:
Thanks @nyh

psarna commented:
@piodul there's a merge conflict between `schema_change_test` and `/next` and I'm not 100% how to resolve it, so please rebase manually and repush.

piodul commented:
@psarna Rebased onto next

psarna commented:
queued as 30d282635
